### PR TITLE
[Enhancement] Bump arrow to 16.1.0

### PR DESCRIPTION
**Pull request by letian-jiang**
## What I'm doing:

By upgrading Arrow, we could leverage the recent community effort including (but not limited to):
- writing parquet files with auxiliary data (page index, bloom filter)
- better to control parquet file-size when general compression algorithm is applied.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by imay** [comment]
@letian-jiang because of Arrow issues, you can split this PR into two separated PRs, one is for zstd and another is for Arrow.

**Comment by letian-jiang** [comment]
> @letian-jiang because of Arrow issues, you can split this PR into two separated PRs, one is for zstd and another is for Arrow.

arrow of version 5.0.0 requires zstd 1.5.0

it looks to me that we cannot upgrade these two libs separately.

https://github.com/apache/arrow/blob/4591d76fce2846a29dac33bf01e9ba0337b118e9/cpp/thirdparty/versions.txt#L59

File: thirdparty/vars.sh [hunk 1]
```diff
@@ -247,10 +247,10 @@ FLATBUFFERS_SOURCE=flatbuffers-1.10.0
 FLATBUFFERS_MD5SUM="f7d19a3f021d93422b0bc287d7148cd2"
 
 # arrow
-ARROW_DOWNLOAD="https://github.com/apache/arrow/archive/apache-arrow-5.0.0.tar.gz"
-ARROW_NAME="arrow-apache-arrow-5.0.0.tar.gz"
-ARROW_SOURCE="arrow-apache-arrow-5.0.0"
-ARROW_MD5SUM="9caf5dbd36ef4972c3a591bcfeaf59c8"
+ARROW_DOWNLOAD="https://github.com/apache/arrow/archive/refs/tags/apache-arrow-16.1.0.tar.gz"
+ARROW_NAME="arrow-apache-arrow-16.1.0.tar.gz"
```
Comment (stephen-shelby) [inline thirdparty/vars.sh:251]: Better run a regression test

**Review by dirtysalt — APPROVED** [review]

File: thirdparty/patches/arrow-16.1.0-use-zstd-1.5.0.patch [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+index 67275ffea..983744990 100644
+--- a/cpp/thirdparty/versions.txt
++++ b/cpp/thirdparty/versions.txt
+@@ -117,8 +117,8 @@ ARROW_XSIMD_BUILD_VERSION=13.0.0
+ ARROW_XSIMD_BUILD_SHA256_CHECKSUM=8bdbbad0c3e7afa38d88d0d484d70a1671a1d8aefff03f4223ab2eb6a41110a3
+ ARROW_ZLIB_BUILD_VERSION=1.3.1
+ ARROW_ZLIB_BUILD_SHA256_CHECKSUM=9a93b2b7dfdac77ceba5a558a580e74667dd6fede4585b91eefb60f03b72df23
+-ARROW_ZSTD_BUILD_VERSION=1.5.6
+-ARROW_ZSTD_BUILD_SHA256_CHECKSUM=8c29e06cf42aacc1eafc4077ae2ec6c6fcb96a626157e0593d5e82a34fd403c1
++ARROW_ZSTD_BUILD_VERSION=1.5.0
```
Comment (kevincai) [inline thirdparty/patches/arrow-16.1.0-use-zstd-1.5.0.patch:11]: image


is it possible to disable this auto zstd download and building, guide arrow to use the zstd version our thirdparty provided?

Comment (letian-jiang) [reply thirdparty/patches/arrow-16.1.0-use-zstd-1.5.0.patch:11]: image
Looks like our building script requires this downloading step.

Comment (letian-jiang) [reply thirdparty/patches/arrow-16.1.0-use-zstd-1.5.0.patch:11]: besides we have provided 
```
export ARROW_ZSTD_URL=${TP_SOURCE_DIR}/${ZSTD_NAME}
```
which direct the build process to read from local file.

**Review by kevincai — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Comment by letian-jiang** [comment]
@merfigy backport branch-3.3

**Comment by derekperkins** [comment]
Downgrading zstd from v1.5.6 to v1.5.0 shouldn't be necessary. Here's the commit for arrow v16.1.0, and it is using zstd v1.5.6
https://github.com/apache/arrow/blob/7dd1d34074af176d9e861a360e135ae57b21cf96/cpp/thirdparty/versions.txt

**Comment by kevincai** [comment]
> Downgrading zstd from v1.5.6 to v1.5.0 shouldn't be necessary. Here's the commit for arrow v16.1.0, and it is using zstd v1.5.6 https://github.com/apache/arrow/blob/7dd1d34074af176d9e861a360e135ae57b21cf96/cpp/thirdparty/versions.txt

It is due to the zstd in sr thirdparty is still v1.5.0, need to keep the dependent library the same version across all libraries.

**Comment by derekperkins** [comment]
I wasn't able to find any references to v1.5.0, so I opened a separate issue for discussion.
https://github.com/StarRocks/starrocks/issues/48393

**Pull request merged**